File: cpp/p14.cc
#include <iostream>
#include <unordered_map>
#include <list>

int max_chain = 0;
int max_key = 0;
std::unordered_map<int, int> collatz_count;

void update_max(int key, int val) {
	if (val > max_chain) {
		max_chain = val;
		max_key = key;
	}
}

void add_collatz(int n) {
	// The natural recursive method for this function does not work because the
	// maximum recursion depth is exceeded.
	std::list<int> values_to_add;
	while (collatz_count.count(n) == 0) {
		values_to_add.push_front(n);
		if (n % 2 == 0) {
			n = n / 2;
		} else {
			n = 3 * n + 1;
		}
	}

	int step = 1;
	std::cout << values_to_add.size() << std::endl;
	for (std::list<int>::iterator it = values_to_add.begin(); 
			it != values_to_add.end(); ++it) {
		collatz_count[*it] = collatz_count[n] + step;
		step++;
	}
	std::cout << values_to_add.size() << std::endl;
}

int main() {
	collatz_count[1] = 0;
	for (int i = 2; i < 200000; i++) {
		add_collatz(i);
		update_max(i, collatz_count[i]);
	}

	std::cout << max_chain << std::endl;
	std::cout << max_key << std::endl;

	return 0;
}

File: p14.py
collatz_count = {1:0}

max_value = 0
max_key = 1

def update_max(k, v):
	global max_value
	global max_key
	if v > max_value:
		max_value = v
		max_key = k

def add_collatz(n):
	if not n in collatz_count:
		if n % 2 == 0:
			add_collatz(n/2)
			collatz_count[n] = collatz_count[n/2] + 1
		else:
			add_collatz(3 * n + 1)
			collatz_count[n] = collatz_count[3 * n + 1] + 1

for i in range(2, 1000000):
	add_collatz(i)
	update_max(i, collatz_count[i])


print max_key